include/libgit4cpp/Remote.h

```diff
@@ -0,0 +1,82 @@
+/**
+ * \file   Remote.h
+ * \author Lars Fröhlich
+ * \date   Created on January 15, 2024
+ * \brief  Declaration of the Remote class.
+ *
+ * \copyright Copyright 2024 Deutsches Elektronen-Synchrotron (DESY), Hamburg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#ifndef LIBGIT4CPP_REMOTE_H_
+#define LIBGIT4CPP_REMOTE_H_
+
+#include <string>
+
+#include <git2.h>
+#include <gul14/optional.h>
+#include <gul14/SmallVector.h>
+
+#include "libgit4cpp/types.h"
+
+namespace git {
+
+/**
+ * An abstraction of a git remote (such as "origin").
+ *
+ * A remote has a name and a URL which can be accessed via get_name() and get_url(),
+ * respectively.
+ */
+class Remote
+{
+public:
+    /**
+     * Construct a Remote object by taking the ownership of a git_remote unique pointer.
+     * The constructor queries the name and URL of the remote and caches them for later
+     * access via get_name() and get_url().
+     * \exception Error is thrown if the given pointer is null.
+     */
+    Remote(LibGitRemote&& remote_ptr);
+
+    /// Return a non-owning pointer to the underlying git remote object.
+    git_remote* get() const { return remote_.get(); }
+
+    /// Return the name of the remote (e.g. "origin"). May be empty for in-memory remotes.
+    const std::string& get_name() const { return name_; }
+
+    /// Return the URL of the remote (e.g. "https://gitlab.com/a/b.git").
+    const std::string& get_url() const { return url_; }
+
+    /**
+     * Retrieve a list of references available on this remote repository
+     * ("git ls-remote").
+     *
+     * \note
+     * The API for this function is preliminary. It would be more useful to return not
+     * just the names of the references, but also additional details like the commit IDs.
+     */
+    std::vector<std::string> list_references();
+
+private:
+    std::string name_;
+    std::string url_;
+    LibGitRemote remote_{ nullptr, git_remote_free };
```

Comment:
> I'm also not sure that Remote should be allowed to outlive the originating GetRepository, because
git::Remote.get()->git_remote_owner() return a pointer to the repository object that might or might not be living anymore.

I see your point, but I do not think we need to address that. IMHO if users start making their own calls into `libgit2`, they are responsible for their own mess. I would see things differently if you couldn't safely call `Remote` member functions anymore once the parent repo has been destructed, but that is not the case. (... unless it is?)

Reply:
The `git_remote` object:

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/aecc252a-d883-4f67-8934-0159d96df39a)

That is filled directly with the same pointer value as out `GitRepository.get()` (but now non-owning), thus is invalid if the repository goes out of scope.

See https://github.com/libgit2/libgit2/blob/25e2b9d8c217e92eb70cde85bb4007a3dec0428d/src/libgit2/remote.c#L502

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/7553bd84-a88c-4e5b-8cbc-5ab122b70e5e)

That seems to be used by the `transport` that is used for example by `git_remote_ls()`, if ssh is used?
Well I did not do an extensive search.

For me it seems a `git_remote` does need the existing `git_repository` and we should make sure of that even if that is more work - is that not what C++ is for in comparison to C, that dangling pointers are avoided?

And further I also think they are responsible for their own mess, but WE should try to prevent any footguns we can; and that is doable for `git::Remote` that know if their respective `git::GitRepository` has been destructed.
If that is too much work I can implement that, did that a couple of times already for similar cases.

Reply:
At least we should try what comes out with asan if
* Create a `GitRepository` for an existing git repo on the filesystem that has remotes
* Crate a `Remote` from that
* Destruct the `GitRepository`
* See if `Remote::listRreferences()` is an asan problem

If yes the need is clear.
If not we can discuss if footgun prevention is needed or documentation is enough.

Reply:
I have added a test for that: 2d06bb77d2596a43acc1b96cde25a41b82070498

No problems with the address sanitizer on Ubuntu 20 LTS. I see no need for further safeguards.

Reply:
Great! Thanks for performing the test.
Then I believe we leave it as it is and possibly improve it later (if at all ;-> )